src/main/scala-2/chisel3/simulator/HasSimulator.scala

```diff
@@ -0,0 +1,54 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.simulator
+
+import java.nio.file.Files
+
+/** Type class for providing a simulator. */
+trait HasSimulator {
+
+  /** Return a simulator.  This require providing a type class implementation of
+    * [[HasTestingDirectory]].
+    */
+  def getSimulator(implicit testingDirectory: HasTestingDirectory): Simulator[_]
+
+}
+
+/** Type class implementations of [[HasSimulator]]. */
+object HasSimulator {
+
+  /** A [[HasSimulator]] implementation for a Verilator simulator. */
+  val verilator: HasSimulator = new HasSimulator {
+    override def getSimulator(implicit testingDirectory: HasTestingDirectory): Simulator[svsim.verilator.Backend] =
+      new Simulator[svsim.verilator.Backend] {
+        override val backend = svsim.verilator.Backend.initializeFromProcessEnvironment()
+        override val tag = "verilator"
+        override val commonCompilationSettings = svsim.CommonCompilationSettings()
+        override val backendSpecificCompilationSettings = svsim.verilator.Backend.CompilationSettings()
+        override val workspacePath = Files.createDirectories(testingDirectory.getDirectory).toString
+      }
+  }
+
+  /** A [[HasSimulator]] implementation for a VCS simulator. */
+  val vcs: HasSimulator = new HasSimulator {
+    override def getSimulator(implicit testingDirectory: HasTestingDirectory): Simulator[svsim.vcs.Backend] =
+      new Simulator[svsim.vcs.Backend] {
+        override val backend = svsim.vcs.Backend.initializeFromProcessEnvironment().getOrElse {
+          throw new Exception(
+            "Unable to load VCS from the environment. (Did you forget to set VCS_HOME and LM_LICENSE_FILE?)"
+          )
+        }
+        override val tag = "vcs"
+        override val commonCompilationSettings = svsim.CommonCompilationSettings()
+        override val backendSpecificCompilationSettings = svsim.vcs.Backend.CompilationSettings()
+        override val workspacePath = Files.createDirectories(testingDirectory.getDirectory).toString
+      }
+  }
+
+  /** Low-priority default implementation of [[HasSimulator]] that uses Verilator.
+    * This is the default that will be used if the user does provide an
+    * alternative.
+    */
+  implicit val default: HasSimulator = verilator
```

Comment:
```suggestion
  implicit def default: HasSimulator = verilator
```
Nit, there's no reason for this to be a val

Reply:
Changed in: a312ee898